ripgrep: fix for linuxbrew

timotheecour opened the pull request:
FYI @sjackman  

- [x] Have you followed the [guidelines for contributing](https://github.com/Linuxbrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Linuxbrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

* doesn't pass audit yet (WIP): Commented-out dep "rust" => :build

* ripgrep depends on `rust => :build` which is broken on linuxbrew (see https://github.com/Linuxbrew/homebrew-core/issues/1733 and https://github.com/Linuxbrew/homebrew-core/pull/1744)

* furthermore, building rust takes 40 minutes to build (https://github.com/Homebrew/homebrew-core/issues/9703), so would be good to allow a user-built version of rust to avoid this slow build, at least as an option

This PR allows using a user defined version of rust (eg installed via https://www.rust-lang.org/en-US/install.html running `curl https://sh.rustup.rs -sSf | sh`). cargo installs ok:
```
which cargo
/home/timothee/.cargo/bin/cargo
```

However, I'm running into another issue:

```
brew install --keep-tmp --verbose ripgrep
==> cargo build --release
error: no default toolchain configured

Temporary files retained at /tmp/ripgrep-20170210-9397-50vknw
```

this is weird because:

```
cd /tmp/ripgrep-20170210-9397-50vknw
cd ripgrep-0.4.0
cargo build --release
# installs rg correctly:
./target/release/rg -h
#works
```

Questions:

* Q1 how do I change the PR to allow depending on rust but not building if cargo is in PATH?

* Q2  how come `cargo build --release` fails in brew but succeeds in the tmp-dir, and how do I fix it? i also tried with `--env=std` or `--env=super` (and passing the full path to cargo) but didn't help

timotheecour commented:
update: figured out why i got `error: no default toolchain configured`: it' because HOME was being reset to the tmp dir. It works if I either do `export RUSTUP_HOME=$HOME/.multirust && brew install ripgrep` or if I hardcode path to my `$HOME` in `system 'export RUSTUP_HOME=/home/timothee/.multirust && cargo build --release` neither of which is great

@sjackman : Question:
* Q3 how should i replace `system 'export RUSTUP_HOME=/home/timothee/.multirust && cargo build --release'` in the PR?

timotheecour commented:
@sjackman 

update: Q3 still holds, but just committed a better change, which works so long user has installed rust and ran `export RUSTUP_HOME=$HOME/.multirust`

timotheecour commented:
filed https://github.com/rust-lang-nursery/rustup.rs/wiki

sjackman commented:
I know little (nothing really) about Rust, so I'm not the best person to answer these questions.

> Q3 how should i replace `system 'export RUSTUP_HOME=/home/timothee/.multirust && cargo build --release'` in the PR?

```ruby
ENV["RUSTUP_HOME"] = ENV["HOME"] + "/.multirust"
system "cargo", "build", "--release"
```

sjackman commented:
Is this PR blocked by `rust` #1744?

rwhogg commented:
> Is this PR blocked by rust #1744?

I merged that, so it shouldn't be. That said, I'm building a bottle in Docker because of timeouts.

rwhogg commented:
@timotheecour does this formula still require changes now that `rust` builds?

I wasn't able to bottle `rust` unfortunately; we're reevaluating the relationship between `gcc` and `binutils` which should hopefully resolve those issues relatively soon.

timotheecour commented:
found and fixed following issue: a previous install of rust caused it to download the wrong file in the cache, making subsequent install fail even after https://github.com/Linuxbrew/homebrew-core/pull/1744#issuecomment-279344655

https://github.com/Linuxbrew/homebrew-core/pull/1780 fixes this side effect by changing the version. but doesn't seem related to not being able to build bottle.

as for this PR: doesn't seem needed anymore now that i fixed rust, assuming we can have a rust bottle, so as to skip the 40mn wait

@rwhogg side questions: 
* on osx there's a standard prefix (/usr/local); but not on linux is that correct? 
* does that mean bottles work regardless of where linuxbrew is installed?

sjackman commented:
> on osx there's a standard prefix (/usr/local); but not on linux is that correct?

On Linuxbrew the standard prefix is `/home/linuxbrew/.linuxbrew`

> does that mean bottles work regardless of where linuxbrew is installed?

Relocatable bottles are determined automatically and flagged with `cellar :any` and can be installed anywhere. Otherwise bottles can only be installed in `/home/linuxbrew/.linuxbrew`

brson commented:
I don't fully understand the homebrew issues here, but if the `brew` command changes the `HOME` variable then that indeed confuses rustup. Having the recipe set `CARGO_HOME` and `RUSTUP_HOME` should fix it (I imagine it is doing something to configure the environment now since brew is finding `~/.cargo/bin`). Setting `RUSTUP_HOME` should not have any impact on non-rustup installs.

brson commented:
Oh and `~/.rustup` is possibly a more correct value. That is the default name of `RUSTUP_HOME` today, though `~/.multirust` remains a symlink for backwards compatibility. `~/.multirust` won't go away any time soon though, so it may make sense to use it for now. Not sure.

`CARGO_HOME` should probably be handled similarly, though the opportunity for weirdness there is less.
Depending on how brew expects to interop with other tools it may be important to not overwrite these two environment variables if they are already set as well.

Right now there's no simple way to have rustup or cargo tell you the definitive name of `RUSTUP_HOME` and `CARGO_HOME` for the current environment, but one could imagine for tools that want to modify `HOME` that it should be possible to capture the current environment's configuration and reuse it.

Another way to approach this that (I think) would work, would be to detect whether rustup exists, and run the command that modifies `HOME` under `rustup run`. `rustup run` will configure the environment to find the correct toolchain independently of `HOME`.

timotheecour commented:
@rwhogg

> @timotheecour does this formula still require changes now that rust builds?

after the pr i sent https://github.com/Linuxbrew/homebrew-core/pull/1744 it builds fine as is. However, this pr https://github.com/Linuxbrew/homebrew-core/pull/1762 helps with running time, avoiding to build rust (40mn) if it can use a preinstalled one.

I'll close despite that.
